program/rust/src/tests/test_upd_aggregate.rs

```diff
@@ -0,0 +1,288 @@
+            price_account.try_borrow_mut_data().unwrap().as_mut_ptr(),
+            1001,
+            4,
+        ));
+    }
+
+    {
+        let price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+
+        assert_eq!(price_data.agg_.price_, 245);
+        assert_eq!(price_data.agg_.conf_, 85);
+        assert_eq!(price_data.twap_.val_, 125);
+        assert_eq!(price_data.twac_.val_, 28);
+        assert_eq!(price_data.num_qt_, 4);
+        assert_eq!(price_data.timestamp_, 4);
+        assert_eq!(price_data.last_slot_, 1001);
+        assert_eq!(price_data.prev_slot_, 1000);
+        assert_eq!(price_data.prev_price_, 200);
+        assert_eq!(price_data.prev_conf_, 90);
+        assert_eq!(price_data.prev_timestamp_, 3);
+    }
+
+    unsafe {
+        assert!(c_upd_aggregate(
+            price_account.try_borrow_mut_data().unwrap().as_mut_ptr(),
+            1025,
+            5,
+        ));
+    }
+
+    {
+        let price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+
+        assert_eq!(price_data.agg_.status_, PC_STATUS_TRADING);
+        assert_eq!(price_data.last_slot_, 1025);
+        assert_eq!(price_data.num_qt_, 4);
+        assert_eq!(price_data.timestamp_, 5);
+        assert_eq!(price_data.prev_slot_, 1001);
+        assert_eq!(price_data.prev_price_, 245);
+        assert_eq!(price_data.prev_conf_, 85);
+        assert_eq!(price_data.prev_timestamp_, 4);
+    }
+
+    // check what happens when nothing publishes for a while
+    unsafe {
+        assert!(!c_upd_aggregate(
+            price_account.try_borrow_mut_data().unwrap().as_mut_ptr(),
+            1026,
+            10,
+        ));
+    }
+    {
+        let price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+
+        assert_eq!(price_data.agg_.status_, PC_STATUS_UNKNOWN);
+        assert_eq!(price_data.last_slot_, 1025);
+        assert_eq!(price_data.num_qt_, 0);
+        assert_eq!(price_data.timestamp_, 10);
+        assert_eq!(price_data.prev_slot_, 1025);
+        assert_eq!(price_data.prev_price_, 245);
+        assert_eq!(price_data.prev_conf_, 85);
+        assert_eq!(price_data.prev_timestamp_, 5);
+    }
+
+    unsafe {
+        assert!(!c_upd_aggregate(
+            price_account.try_borrow_mut_data().unwrap().as_mut_ptr(),
+            1028,
+            12,
+        ));
+    }
+
+    {
+        let price_data = load_checked::<pc_price_t>(&price_account, PC_VERSION).unwrap();
+
+        assert_eq!(price_data.agg_.status_, PC_STATUS_UNKNOWN);
+        assert_eq!(price_data.last_slot_, 1025);
+        assert_eq!(price_data.num_qt_, 0);
+        assert_eq!(price_data.timestamp_, 12);
+        assert_eq!(price_data.prev_slot_, 1025);
+        assert_eq!(price_data.prev_price_, 245);
+        assert_eq!(price_data.prev_conf_, 85);
+        assert_eq!(price_data.prev_timestamp_, 5);
+    }
+}
+
+// Create an upd_price instruction that sets the product metadata to strings
```

Comment:
comment is wrong

also why Option here for the price/conf/pub_slot?

Reply:
Ok